### PR TITLE
Issue 14534: Lower non-scalar types valid in a boolean context in the front-end

### PR DESCRIPTION
**Pull request by ibuclaw**
As per: https://issues.dlang.org/show_bug.cgi?id=14534 - does the lowering of non-scalar boolean conditions allowed in D to valid scalar values.

This may be a deviation from what GDC/LDC/DMD _actually_ does.  But seems to correlate with what people _expected_.  ie: People were generally surprised when I mentioned that GDC lowers `if (arr)` and `if (dg)` to `(arr.ptr | arr.length)` and `(dg.ptr | dg.funcptr)` respectively.  Though in the edge case where an expression is evaluated true where the user might expect false, there is some serious circumvention going on at runtime.

This bypasses going through DotIdExp - I did have a check and couldn't see any stand out reason to do so.

### CODE REVIEW

File: src/expression.c [hunk 1]
```diff
@@ -2761,6 +2761,29 @@ Expression *Expression::toBoolean(Scope *sc)
             goto Lagain;
         }
     }
+    // Other non-scalar types that are valid for bool conditions
+    // should have their contents lowered to valid scalars.
+    else if (tb->ty == Tarray)
```
Comment (ibuclaw) [inline src/expression.c]: I take it that the auto-tester doesn't like comments before ``else if`` conditions?

Comment (yebblies) [inline src/expression.c]: Magicport doesn't like them.

**Comment by yebblies** [comment]
`(arr.ptr | arr.length)` wait really?  This is a good argument in favor of fixing 4733 by changing the behavior.

**Comment by ibuclaw** [comment]
> (arr.ptr | arr.length) wait really? 

Yes, and it looks like DMD does the same too, see here: https://github.com/ibuclaw/d-programming-language.org/commit/9299d7ddfb82b5406f753285117b78a8a2d217f9#commitcomment-10912398

Please correct me if I am mistaken here:
```
mov    QWORD PTR [rbp-0x10],rax  // move arr.length into rax
mov    QWORD PTR [rbp-0x8],rdx   // move arr.ptr into rdx
or     rdx,rax                   // Test 'arr.ptr | arr.length'
```

**Comment by yebblies** [comment]
Oh right, the problem is that if the length is zero we shouldn't care about the pointer.
```
val         ptr     length  if      !is null    != null
null        null    0       false   false       false
arr[$..$]   !null   0       true    true        false
arr         !null   !0      true    true        true
null[a..b]  null    !0      true    true        true
```
Gah.  If only we could just change if to != null...

**Comment by ibuclaw** [comment]
> Oh right, the problem is that if the length is zero we shouldn't care about the pointer.

In practice yes.  But as I understood it, those who 'claimed' that the ``if(arr)`` warning was bringing up too many false positives were saying so because they didn't care about the length - regardless of zero or otherwise?

**Comment by yebblies** [comment]
I think it was mostly that `null` was used as a special marker value.

File: src/expression.c (hunk 1 above)
Comment (ibuclaw) [inline src/expression.c]: Pity. It doesn't make sense to put the comment anywhere else.

I'll just remove the ``else`` then...

Comment (yebblies) [inline src/expression.c]: You could always update magicport to accept it properly, the source is in master!

**Comment by yebblies** [comment]
Yeeeah...  As I said in the bug report, I think this lowering is too early.  It should be done right before e2ir, if at all.  Lowering early does help avoid code duplication but it's not always worth the control you lose.

**Comment by ibuclaw** [comment]
Code is right, CTFE is wrong.  This is a fine example:

```
assert([5 : 4].dup);  // Error: pointer cast from int[int] to void* is not supported at compile time
```

It must support *some* type of conversion to pointer, unless it uses some brute force method that is akin to having two distinct types for AA's (similar to the mess we used to have in the backend).

**Comment by yebblies** [comment]
I have no idea what you mean.  CTFE does not need to support conversion from `V[K]` to `void*` in order to test an AA for null-ness.

**Comment by ibuclaw** [comment]
So the way I see it, ``AssocArrayLiteralExp::isBool`` can make CTFE produce a different result at compile time vs. runtime.

**Comment by ibuclaw** [comment]
Example:
```
int[int] aa = [5 : 4];
aa.remove(5);
if (__ctfe)
    assert(!aa);  // Passes, CTFE checks keys.dim
else
    assert(aa);  // Passes, runtime only checks .ptr
```

**Comment by ibuclaw** [comment]
I've lost interest in all D compilers doing the same thing for now...
